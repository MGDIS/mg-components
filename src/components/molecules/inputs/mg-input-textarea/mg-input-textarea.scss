@use '../styles/mg-input.scss';
@use '../styles/mg-input-box.scss';
@use '../styles/mg-input-width.scss';
@use '../styles/mg-input-charleft.scss';
@use '../../../../styles/a11y.scss';

/* Component Custom Style */
.mg-input {
  .mg-input__box {
    height: auto;
  }
}

/* Firefox selector */
@-moz-document url-prefix() {
  .mg-input__box {
    overflow-x: hidden; // remove Firefox bottom "phantom scrollbar" in textarea 
  }
}
