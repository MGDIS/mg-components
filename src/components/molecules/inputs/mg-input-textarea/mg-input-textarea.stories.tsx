import { h } from "@stencil/core";

export default {
  component: 'mg-input-textarea',
  title: 'Molecules/Inputs/mg-input-textarea',
};

/**
 * 1. camelCase arguments must be written in the template, for exemple labelOnTop must be placed in the template as label-on-top={args.labelOnTop}
 * 2. boolean arguments with a default true value must be added like display-character-left={args.displayCharacterLeft ? 'true' : 'false'}
 */
 const Template = args => {
  const labelOnTop = args.labelOnTop;
  delete args.labelOnTop;
  const labelHide = args.labelHide;
  delete args.labelHide;
  const patternErrorMessage = args.patternErrorMessage;
  delete args.patternErrorMessage;
  const displayCharacterLeft = args.displayCharacterLeft;
  delete args.displayCharacterLeft;
  const characterLeftTemplate = args.characterLeftTemplate;
  delete args.characterLeftTemplate;
  const helpText = args.helpText;
  delete args.helpText;
  // return element
  return <mg-input-textarea
    {...args}
    label-on-top={labelOnTop}
    label-hide={labelHide}
    pattern-error-message={patternErrorMessage}
    display-character-left={displayCharacterLeft}
    character-left-template={characterLeftTemplate}
    help-text={helpText}
  ></mg-input-textarea>
};

export const MgInputTextarea = Template.bind({});
MgInputTextarea.args = {
  // Global
  value: '',
  identifier: 'identifier',
  name: 'input-name',
  // Label
  label: 'Label',
  labelOnTop: false,
  labelHide: false,
  // Input
  placeholder: 'placeholder',
  maxlength: 4000,
  required: true,
  disabled: false,
  readonly: false,
  pattern: undefined,
  patternErrorMessage: undefined,
  rows: 3,
  // Tooltip
  tooltip: 'This is a tooltip',
  // Nb Char Left
  displayCharacterLeft: true,
  characterLeftTemplate: '{counter} characters left.',
  // Help Text
  helpText: 'Help text with html <strong>bold</strong>, <em>italic</em>.',
};
