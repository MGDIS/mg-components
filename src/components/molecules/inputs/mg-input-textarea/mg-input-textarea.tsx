import { Component, Event, h, Prop, EventEmitter, State } from '@stencil/core';
import { MgInput } from '../MgInput';
import { createID } from '../../../../utils/utils';
import locale from '../../../../locales';

@Component({
  tag: 'mg-input-textarea',
  styleUrl: 'mg-input-textarea.scss',
  shadow: true,
})
export class MgInputTextarea {

  /************
   * Internal *
   ************/

   private classFocus = 'is-focused';
   private classError = 'is-not-valid';

  /**************
  * Decorators *
  **************/

  /**
   * Component value
   */
  @Prop({ mutable:true, reflect: true }) value: string;

  /**
   * Identifier is used for the element ID (id is a reserved prop in Stencil.js)
   * If not set, it will be created.
   */
  @Prop() identifier?: string = createID('mg-input-textarea');

  /**
   * Input name
   * If not set the value equals the identifier
   */
  @Prop() name?: string = this.identifier;

  /**
   * Input label
   * Required
   */
  @Prop() label!: string;

  /**
   * Define if label is displayed on top
   */
   @Prop() labelOnTop: boolean;

  /**
   * Define if label has colon ":"
   */
  @Prop() labelColon: boolean = false;

  /**
   * Input placeholder.
   * It should be a word or short phrase that demonstrates the expected type of data, not a replacement for labels or help text.
   */
  @Prop() placeholder: string;

  /**
   * Input max length
   */
  @Prop() maxlength: number = 4000;

  /**
   * Define if input is required
   */
  @Prop() required: boolean = false;

  /**
   * Define if input is readonly
   */
  @Prop() readonly: boolean = false;

  /**
   * Define if input is disabled
   */
  @Prop() disabled: boolean = false;

  /**
   * Define input pattern to validate
   */
  @Prop() pattern: string;

  /**
   * Define input pattern error message
   */
  @Prop() patternErrorMessage: string;

  /**
   * Define input pattern error message
   */
   @Prop() rows: number = 3;

  /**
   * Add a tooltip message next to the input
   */
   @Prop() tooltip: string;

  /**
   * Define if component should display character left
   */
  @Prop() displayCharacterLeft: boolean = true;

  /**
   * Template to use for characters left sentence
   */
  @Prop() characterLeftTemplate: string;

  /**
   * Template to use for characters left sentence
   */
  @Prop() helpText: string;

  /**
   * Define input pattern to validate
   */
  @Prop({ mutable: true, reflect: true }) valid: boolean;

  /**
  * Define input pattern error message
  */
  @Prop({ mutable: true, reflect: true }) invalid: boolean;

  /**
   * Component classes
   */
  @State() classes: Set<string> = new Set(['mg-input--textarea']);

  /**
   * Error message to display
   */
  @State() errorMessage: string;

  /**
   * Emmited event when value change
   */
  @Event() valueChange: EventEmitter<string>

  /**
   * Handle input event
   * @param event
   */
   private handleInput = (event:InputEvent & { target: HTMLInputElement }) => {
    this.value = event.target.value;
    this.valueChange.emit(this.value);
  }

  /**
   * Handle focus event
   */
  private handleFocus = () => {
    this.classes.add(this.classFocus)
    this.classes = new Set(this.classes);
  }

  /**
   * Handle blur event
   * @param event
   */
  private handleBlur = (event:FocusEvent) => {
    // Manage focus
    this.classes.delete(this.classFocus);
    this.classes = new Set(this.classes);
    // Check validity
    this.checkValidity(event.target);
  }

  /**
   * Check if input is valid
   * @param element
   */
  private checkValidity(element) {
    // Pattern is not defined on textarea field : https://developer.mozilla.org/fr/docs/Web/HTML/Element/Textarea
    const patternValidity = this.pattern === undefined || new RegExp(`^${this.pattern}$`, 'u').test(this.value);
    const validity = element.checkValidity() && patternValidity;
    // Set error message
    this.errorMessage = undefined;
    if(!validity && !patternValidity){
      this.errorMessage = this.patternErrorMessage;
    }
    // required
    else if(!validity && element.validity.valueMissing) {
      this.errorMessage = locale.errors.required;
    }

    // Set validity
    this.valid = validity;
    this.invalid = !validity;

    // Update class
    if(validity) {
      this.classes.delete(this.classError);
      this.classes = new Set(this.classes);
    }
    else {
      this.classes.add(this.classError);
      this.classes = new Set(this.classes);
    }
  }

  /**
   * Validate patern configuration
   */
  private validatePattern() {
    if(
      this.pattern && typeof this.pattern === 'string' && this.pattern !== '' &&
      (this.patternErrorMessage === undefined || typeof this.patternErrorMessage !== 'string' || this.patternErrorMessage === '')
    ) {
      throw new Error('<mg-input-textarea> prop "pattern" must be paired with the prop "patternErrorMessage"')
    }
  }

  /*************
   * Lifecycle *
   *************/

  /**
   * Check if component props are well configured on init
   */
  componentWillLoad() {
    this.validatePattern();
  }

  render() {
    return (
      <MgInput
        identifier={this.identifier}
        classes={this.classes}
        label={this.label}
        labelOnTop={this.labelOnTop}
        labelColon={this.labelColon}
        required={this.required}
        readonly={this.readonly}
        value={this.value}
        readonlyValue={undefined}
        tooltip={this.tooltip}
        displayCharacterLeft={this.displayCharacterLeft}
        characterLeftTemplate={this.characterLeftTemplate}
        maxlength={this.maxlength}
        helpText={this.helpText}
        errorMessage={this.errorMessage}
      >
        <textarea
          id={this.identifier}
          name={this.name}
          placeholder={this.placeholder}
          title={this.placeholder}
          rows={this.rows}
          maxlength={this.maxlength}
          disabled={this.disabled}
          required={this.required}
          onInput={this.handleInput}
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        >{this.value}</textarea>
      </MgInput>
    );
  }
}
