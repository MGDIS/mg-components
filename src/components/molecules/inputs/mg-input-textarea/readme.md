## Design

### Indication du nombre de caractères restants

- lorsque le focus est sur le champ de saisie, un message s'affiche en dessous
- le message est : "<nbCar> caractères disponibles", <nbCar> étant un compteur qui se met à jour dynamiquement en fonction de la saisie
- lorsque le focus n'est plus sur le champ de saisie, le message disparaît

### Valeur

- par défaut limitée 4000 caractères de type alpha numérique.

### Dimensions

- la hauteur du composant est par défaut de 3 lignes de texte (cette valeur est configurable)
- le champ est retaillable en cliquant en bas à droite de la zone de saisie

<!-- Auto Generated Below -->


## Properties

| Property                | Attribute                 | Description                                                                                                                                    | Type                     | Default                         |
| ----------------------- | ------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------ | ------------------------------- |
| `characterLeftTemplate` | `character-left-template` | Template to use for characters left sentence                                                                                                   | `string`                 | `undefined`                     |
| `disabled`              | `disabled`                | Define if input is disabled                                                                                                                    | `boolean`                | `false`                         |
| `displayCharacterLeft`  | `display-character-left`  | Define if component should display character left                                                                                              | `boolean`                | `true`                          |
| `helpText`              | `help-text`               | Add a help text under the input, usually expected data format and example                                                                      | `string`                 | `undefined`                     |
| `identifier`            | `identifier`              | Identifier is used for the element ID (id is a reserved prop in Stencil.js) If not set, it will be created.                                    | `string`                 | `createID('mg-input-textarea')` |
| `invalid`               | `invalid`                 | Define input invalid state                                                                                                                     | `boolean`                | `undefined`                     |
| `label` _(required)_    | `label`                   | Input label                                                                                                                                    | `string`                 | `undefined`                     |
| `labelHide`             | `label-hide`              | Define if label is visible                                                                                                                     | `boolean`                | `false`                         |
| `labelOnTop`            | `label-on-top`            | Define if label is displayed on top                                                                                                            | `boolean`                | `undefined`                     |
| `maxlength`             | `maxlength`               | Input max length                                                                                                                               | `number`                 | `4000`                          |
| `name`                  | `name`                    | Input name If not set the value equals the identifier                                                                                          | `string`                 | `this.identifier`               |
| `pattern`               | `pattern`                 | Define input pattern to validate                                                                                                               | `string`                 | `undefined`                     |
| `patternErrorMessage`   | `pattern-error-message`   | Define input pattern error message                                                                                                             | `string`                 | `undefined`                     |
| `placeholder`           | `placeholder`             | Input placeholder. It should be a word or short phrase that demonstrates the expected type of data, not a replacement for labels or help text. | `string`                 | `undefined`                     |
| `readonly`              | `readonly`                | Define if input is readonly                                                                                                                    | `boolean`                | `false`                         |
| `required`              | `required`                | Define if input is required                                                                                                                    | `boolean`                | `false`                         |
| `rows`                  | `rows`                    | Define input pattern error message                                                                                                             | `number`                 | `3`                             |
| `tooltip`               | `tooltip`                 | Add a tooltip message next to the input                                                                                                        | `string`                 | `undefined`                     |
| `valid`                 | `valid`                   | Define input valid state                                                                                                                       | `boolean`                | `undefined`                     |
| `value`                 | `value`                   | Component value                                                                                                                                | `string`                 | `undefined`                     |
| `width`                 | `width`                   | Define input width                                                                                                                             | `"full" \| 16 \| 2 \| 4` | `'full'`                        |


## Events

| Event          | Description                         | Type                   |
| -------------- | ----------------------------------- | ---------------------- |
| `input-valid`  | Emited event when checking validity | `CustomEvent<boolean>` |
| `value-change` | Emited event when value change      | `CustomEvent<string>`  |


## Methods

### `displayError() => Promise<void>`

Public method to display errors

#### Returns

Type: `Promise<void>`




## Dependencies

### Depends on

- [mg-tooltip](../../../atoms/mg-tooltip)
- [mg-icon](../../../atoms/mg-icon)
- [mg-input-title](../../../atoms/mg-input-title)
- [mg-character-left](../../../atoms/mg-character-left)

### Graph
```mermaid
graph TD;
  mg-input-textarea --> mg-tooltip
  mg-input-textarea --> mg-icon
  mg-input-textarea --> mg-input-title
  mg-input-textarea --> mg-character-left
  style mg-input-textarea fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
