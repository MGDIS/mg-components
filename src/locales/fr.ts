const locale = {
  "errors": {
    "required": "Ce champs est obligatoire.",
    "date": {
      "badInput": "Le format de la date doit être du type <span aria-hidden=\"true\">jj/mm/aaaa</span><span class=\"sr-only\">j j / m m / a a a a</span> et supérieure au 01/01/1900"
    }
  },
  "nbCharLeft" : "{counter} caractères disponibles.",
  "input": {
    "select" : {
      "placeholder": "Veuillez sélectionner une valeur"
    },
    "checkbox": {
      "true": "Oui",
      "false": "Non",
      "indeterminate": "Indéterminé",
    }
  }
}

export default locale;
