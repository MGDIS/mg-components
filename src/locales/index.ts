import en from './en/messages.json';
import fr from './fr/messages.json';

const defaultLocale = 'en';
const messages = { en, fr };

/**
 * Get locale
 *
 * @param {HTMLElement} element element that needs to know the locale to use
 * @returns {string} locale
 */
const getLocale = (element: HTMLElement): string => {
  return (element.closest('[lang]') as HTMLElement)?.lang || navigator.language || defaultLocale;
};

/**
 * Get Intl object
 * We load the defined locale but for now we only support the first subtag for messages
 *
 * @param {HTMLElement} element element we need to get the language
 * @returns {{ locale: string; messages: unknown }} messages object
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const initLocales = (element: HTMLElement): { locale: string; messages: any } => {
  // Get local
  const locale = getLocale(element);
  // Only keep first subtag
  const localeSubtag = locale.split('-').shift();
  // Return
  return {
    locale,
    messages: messages[localeSubtag] || messages[defaultLocale],
  };
};
