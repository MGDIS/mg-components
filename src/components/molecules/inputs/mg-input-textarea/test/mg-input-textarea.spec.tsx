import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgInputTextarea } from '../mg-input-textarea';
import { messages } from '../../../../../locales';

const getPage = args =>
  newSpecPage({
    components: [MgInputTextarea],
    template: () => <mg-input-textarea {...args}></mg-input-textarea>,
  });

describe('mg-input-textarea', () => {
  test.each([
    { label: 'label', identifier: 'identifier' },
    { label: 'label', identifier: 'identifier', labelHide: true },
    { label: 'label', identifier: 'identifier', labelOnTop: true },
    { label: 'label', identifier: 'identifier', readonly: true },
    { label: 'label', identifier: 'identifier', readonly: true, labelOnTop: true, tooltip: 'Tooltip message' },
    { label: 'label', identifier: 'identifier', readonly: true, value: 'blu' },
    { label: 'label', identifier: 'identifier', tooltip: 'My Tooltip Message' },
    { label: 'label', identifier: 'identifier', tooltip: 'My Tooltip Message', labelOnTop: true },
  ])('Should render with args %s:', async args => {
    const { root } = await getPage(args);
    expect(root).toMatchSnapshot();
  });

  test.each(['', undefined])('Should throw error with invalid label property : %s', async value => {
    try {
      await getPage({ label: value });
    } catch (err) {
      expect(err.message).toMatch('<mg-input> prop "label" is required');
    }
  });

  test('Should throw an error with labelOnTop & labelHide set to true', async () => {
    try {
      await getPage({ label: 'batman', labelOnTop: true, labelHide: true });
    } catch (err) {
      expect(err.message).toMatch('<mg-input> prop "labelOnTop" must not be paired with the prop "labelHide"');
    }
  });

  test.each(['', undefined])('Should throw an error when pattern is used with patternErrorMessage: %s', async value => {
    try {
      const { root } = await getPage({ label: 'blu', pattern: '[a-z]*', patternErrorMessage: value });
      expect(root).toMatchSnapshot();
    } catch (err) {
      expect(err.message).toMatch('<mg-input-textarea> prop "pattern" must be paired with the prop "patternErrorMessage"');
    }
  });

  test('Should trigger events', async () => {
    const inputValue = 'Blu';
    const args = { label: 'label', identifier: 'identifier', helpText: 'My help text' };
    const page = await getPage(args);

    const element = page.doc.querySelector('mg-input-textarea');
    const input = element.shadowRoot.querySelector('textarea');

    //mock validity
    input.checkValidity = jest.fn(() => true);
    Object.defineProperty(input, 'validity', {
      get: jest.fn(() => ({
        valueMissing: false,
      })),
    });

    jest.spyOn(page.rootInstance.valueChange, 'emit');

    input.dispatchEvent(new CustomEvent('focus', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.classList.has('is-focused')).toBeTruthy();

    expect(page.root).toMatchSnapshot(); //Snapshot on focus

    input.value = inputValue;
    input.dispatchEvent(new CustomEvent('input', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.valueChange.emit).toHaveBeenCalledWith(inputValue);

    input.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.classList.has('is-focused')).toBeFalsy();
  });

  test.each([
    { validity: true, valueMissing: false },
    { validity: false, valueMissing: true },
    { validity: false, valueMissing: false, value: 'Blu', pattern: '[a-z]*', patternErrorMessage: 'Non' },
  ])('validity (%s), valueMissing (%s)', async ({ validity, valueMissing, value, pattern, patternErrorMessage }) => {
    const args = { label: 'label', identifier: 'identifier', value, pattern, patternErrorMessage };
    const page = await getPage(args);

    const element = page.doc.querySelector('mg-input-textarea');
    const input = element.shadowRoot.querySelector('textarea');

    //mock validity
    input.checkValidity = jest.fn(() => validity);
    Object.defineProperty(input, 'validity', {
      get: jest.fn(() => ({
        valueMissing,
      })),
    });

    input.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
    await page.waitForChanges();

    if (validity) {
      expect(page.rootInstance.errorMessage).toBeUndefined();
    } else if (valueMissing) {
      expect(page.rootInstance.errorMessage).toEqual(messages.errors.required);
    } else if (pattern !== undefined) {
      expect(page.rootInstance.errorMessage).toEqual(patternErrorMessage);
    }
    expect(page.rootInstance.valid).toEqual(validity);
    expect(page.rootInstance.invalid).toEqual(!validity);
  });
});
