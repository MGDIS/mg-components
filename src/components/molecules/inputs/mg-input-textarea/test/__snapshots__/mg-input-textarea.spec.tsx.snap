// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-input-textarea Should render with args {
  label: 'label',
  identifier: 'identifier',
  readonly: true,
  labelOnTop: true,
  tooltip: 'Tooltip message'
}: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--label-on-top mg-input--readonly mg-input--textarea mg-input--width-full">
      <div class="mg-input__title">
        <mg-input-title identifier="identifier">
          label
        </mg-input-title>
      </div>
      <div class="mg-input__input-container">
        <strong></strong>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args {
  label: 'label',
  identifier: 'identifier',
  readonly: true,
  value: 'blu'
}: 1`] = `
<mg-input-textarea value="blu">
  <mock:shadow-root>
    <div class="mg-input mg-input--readonly mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <strong>
          blu
        </strong>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args {
  label: 'label',
  identifier: 'identifier',
  tooltip: 'My Tooltip Message'
}: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
          <mg-tooltip identifier="identifier-tooltip" message="My Tooltip Message">
            <mg-icon icon="info-circle"></mg-icon>
          </mg-tooltip>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args {
  label: 'label',
  identifier: 'identifier',
  tooltip: 'My Tooltip Message',
  labelOnTop: true
}: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--label-on-top mg-input--textarea mg-input--width-full">
      <div class="mg-input__title">
        <mg-input-title identifier="identifier">
          label
        </mg-input-title>
        <mg-tooltip identifier="identifier-tooltip" message="My Tooltip Message">
          <mg-icon icon="info-circle"></mg-icon>
        </mg-tooltip>
      </div>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args { label: 'label', identifier: 'identifier' }: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args { label: 'label', identifier: 'identifier', labelHide: true }: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title class="sr-only" identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args { label: 'label', identifier: 'identifier', labelOnTop: true }: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--label-on-top mg-input--textarea mg-input--width-full">
      <div class="mg-input__title">
        <mg-input-title identifier="identifier">
          label
        </mg-input-title>
      </div>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args { label: 'label', identifier: 'identifier', readonly: true }: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--readonly mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <strong></strong>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args { label: 'label', identifier: 'identifier', resizable: 'both' }: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="" class="mg-input__box mg-input__box--resizable" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args { label: 'label', identifier: 'identifier', resizable: 'horizontal' }: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="" class="mg-input__box mg-input__box--resizable-horizontal" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should render with args { label: 'label', identifier: 'identifier', resizable: 'vertical' }: 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="" class="mg-input__box mg-input__box--resizable-vertical" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea Should trigger events 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="is-focused mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="identifier-character-left identifier-help-text" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
        <mg-character-left identifier="identifier-character-left" maxlength="4000"></mg-character-left>
        <div class="mg-input__help-text" id="identifier-help-text">
          My help text
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea display error with displayError component's public method 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier" required="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" required="" rows="3"></textarea>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea display error with displayError component's public method 2`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="is-not-valid mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier" required="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="identifier-error" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" required="" rows="3"></textarea>
        </div>
        <div aria-live="assertive" class="mg-input__error" id="identifier-error">
          Ce champ est obligatoire.
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea validity, case next state is disabled validity ({ validity: false, valueMissing: true }), valueMissing (%s) 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="is-not-valid mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="identifier-error" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
        <div aria-live="assertive" class="mg-input__error" id="identifier-error">
          Ce champ est obligatoire.
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea validity, case next state is disabled validity ({ validity: false, valueMissing: true }), valueMissing (%s) 2`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="identifier-error" class="mg-input__box" disabled="" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
        <div aria-live="assertive" class="mg-input__error" id="identifier-error">
          Ce champ est obligatoire.
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea validity, case next state is readonly validity ({ validity: false, valueMissing: true }), valueMissing (%s) 1`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="is-not-valid mg-input mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input"><textarea aria-describedby="identifier-error" class="mg-input__box" id="identifier" maxlength="4000" name="identifier" rows="3"></textarea>
        </div>
        <div aria-live="assertive" class="mg-input__error" id="identifier-error">
          Ce champ est obligatoire.
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;

exports[`mg-input-textarea validity, case next state is readonly validity ({ validity: false, valueMissing: true }), valueMissing (%s) 2`] = `
<mg-input-textarea>
  <mock:shadow-root>
    <div class="mg-input mg-input--readonly mg-input--textarea mg-input--width-full">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <strong></strong>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-textarea>
`;
